config/management/src/verify.rs

```diff
@@ -110,7 +146,134 @@ fn write_waypoint(storage: &dyn Storage, buffer: &mut String, key: &str) {
                     .unwrap_or_else(|_| "Invalid waypoint".into())
             }
         })
-        .unwrap_or_else(|e| format!("{:?}", e));
+        .unwrap_or_else(|e| e.to_string());
 
     writeln!(buffer, "{} - {}", key, value).unwrap();
 }
+
+fn compare_genesis(
+    storage: &dyn Storage,
+    buffer: &mut String,
+    genesis_path: &PathBuf,
+) -> Result<(), Error> {
+    let db_path = TempPath::new();
+    let (db_rw, expected_waypoint) = compute_genesis(genesis_path, db_path.path())?;
+    let validator_account = validator_account(storage)?;
+    let validator_config = validator_config(validator_account, db_rw.reader.clone())?;
+
+    let actual_consensus_key = ed25519_from_storage(CONSENSUS_KEY, storage)
+        .map_err(|e| Error::LocalStorageReadError(CONSENSUS_KEY, e))?;
+    let expected_consensus_key = validator_config.consensus_public_key;
+    write_assert(
+        buffer,
+        CONSENSUS_KEY,
+        actual_consensus_key == expected_consensus_key,
+    );
+
+    let actual_validator_key = x25519_from_storage(VALIDATOR_NETWORK_KEY, storage)
+        .map_err(|e| Error::LocalStorageReadError(VALIDATOR_NETWORK_KEY, e))?;
+    let expected_validator_key = validator_config.validator_network_identity_public_key;
+    write_assert(
+        buffer,
+        VALIDATOR_NETWORK_KEY,
+        actual_validator_key == expected_validator_key,
+    );
+
+    let actual_fullnode_key = x25519_from_storage(FULLNODE_NETWORK_KEY, storage)
+        .map_err(|e| Error::LocalStorageReadError(VALIDATOR_NETWORK_KEY, e))?;
+    let expected_fullnode_key = validator_config.full_node_network_identity_public_key;
+    write_assert(
+        buffer,
+        FULLNODE_NETWORK_KEY,
+        actual_fullnode_key == expected_fullnode_key,
+    );
+
+    let actual_waypoint = storage
+        .get(WAYPOINT)
+        .and_then(|c| c.value.string())
+        .map_err(|e| Error::LocalStorageReadError(WAYPOINT, e.to_string()))?;
+    let actual_waypoint = Waypoint::from_str(&actual_waypoint).map_err(|e| {
+        Error::UnexpectedError(format!("Unable to parse waypoint: {}", e.to_string()))
+    })?;
+    write_assert(
+        buffer,
+        libra_global_constants::WAYPOINT,
+        actual_waypoint == expected_waypoint,
+    );
+
+    Ok(())
+}
+
+fn compute_genesis(
```

Comment:
nit: this method looks more like it computes a waypoint and just reads genesis from the file? perhaps rename to "compute_waypoint"?

Reply:
The DbReaderWriter contains the computation of genesis, I was wondering if I should leave comments, this indicates I should :)